internal/lido/contracts/csfeedistributor/rewards.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	treeCID, err := treeCID(network)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call treeCID: %w", err)
+	}
+
+	shares, err := cumulativeFeeShares(treeCID, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call cumulativeFeeShares: %w", err)
+	}
+
+	proof, err := convertToByte(proofStrings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call convToByte: %w", err)
+	}
+
+	fees, err := contract.GetFeesToDistribute(nil, nodeID, shares, proof)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call GetFeesToDistribute: %w", err)
+	}
+
+	bondInfo, err := bond.BondSummary(network, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call BondSummary: %w", err)
+	}
+
+	rewards = new(big.Int).Add(bondInfo.Excess, fees)
+	return rewards, nil
+}
+
+func cumulativeFeeShares(treeCID string, nodeID *big.Int) (*big.Int, error) {
+	// Get tree data from IPFS
+	treeData, err := treeData(treeCID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call treeData: %v", err)
+	}
+
+	// Compare nodeOperatorID in tree with nodeId to get shares
+	// Binary search for the nodeOperatorId
+	low, high := 0, len(treeData.Values)-1
+	for low <= high {
+		mid := (low + high) / 2
+		nodeOperatorId, err := convertToBigInt(treeData.Values[mid].Value[0])
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert nodeOperatorId: %v", err)
+		}
+		cmp := nodeOperatorId.Cmp(nodeID)
+		if cmp == 0 {
+			// Node operator ID matches, return the shares
+			shares, err := convertToBigInt(treeData.Values[mid].Value[1])
+			if err != nil {
+				return nil, fmt.Errorf("failed to convert shares: %v", err)
+			}
+			return shares, nil
+		} else if cmp < 0 {
+			low = mid + 1
+		} else {
+			high = mid - 1
+		}
+	}
+	return nil, fmt.Errorf("invalid nodeId")
+}
+
+func treeCID(network string) (string, error) {
+	var treeCIDString string
+	contract, err := csFeeDistributorContract(network)
+	if err != nil {
+		return treeCIDString, fmt.Errorf("failed to call csFeeDistributorContract: %w", err)
+	}
+
+	treeCIDString, err = contract.TreeCid(nil)
+	if err != nil {
+		return treeCIDString, fmt.Errorf("failed to call TreeCid: %w", err)
+	}
+	return treeCIDString, nil
+}
+
+func treeData(treeCID string) (Tree, error) {
+	var treeData Tree
+	gatewayURL := fmt.Sprintf("https://ipfs.io/ipfs/%s", treeCID) // Public gateway URL
+	resp, err := http.Get(gatewayURL)
+	if err != nil {
+		return treeData, fmt.Errorf("failed to connect to gatewayURL: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Read and print the data
+	data, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return treeData, fmt.Errorf("failed to read data: %w", err)
+	}
+
+	// Collect data in treeData struct
+	if err := json.Unmarshal(data, &treeData); err != nil {
+		return treeData, fmt.Errorf("failed to unmarshal JSON: %w", err)
+	}
+	return treeData, nil
+}
+
+func convertToBigInt(value interface{}) (*big.Int, error) {
```

Comment:
Convert what to BigInt? Unit test this please

Reply:
[Done](https://github.com/NethermindEth/sedge/pull/401/commits/d8704a467fe0ee9f41517706e62059f2ffb545c2)

Reply:
I think what @AntiD2ta suggested, apart from unit tests, was that we should have in the name what we are trying to convert to BigInt

Reply:
Adjusted in [this commit](https://github.com/NethermindEth/sedge/pull/401/commits/27f7757898e1b136611fd13fcdd4d5bc8fcb39ce)